libs/langchain/langchain/vectorstores/mongodb_atlas.py

```diff
@@ -159,24 +198,64 @@ def _similarity_search_with_score(
         }
         if pre_filter:
             knn_beta["filter"] = pre_filter
-        pipeline = [
-            {
-                "$search": {
-                    "index": self._index_name,
-                    "knnBeta": knn_beta,
-                }
-            },
-            {"$set": {"score": {"$meta": "searchScore"}}},
-        ]
-        if post_filter_pipeline is not None:
-            pipeline.extend(post_filter_pipeline)
-        cursor = self._collection.aggregate(pipeline)  # type: ignore[arg-type]
-        docs = []
-        for res in cursor:
-            text = res.pop(self._text_key)
-            score = res.pop("score")
-            docs.append((Document(page_content=text, metadata=res), score))
-        return docs
+        query = {
+            "$search": {
+                "index": self._index_name,
+                "knnBeta": knn_beta,
+            }
+        }
+
+        return self._similarity_search_query(query, post_filter_pipeline)
+
+    def _similarity_search_query_vectorsearch(
+        self,
+        embedding: List[float],
+        k: int = 4,
+        pre_filter: Optional[dict] = None,
+        post_filter_pipeline: Optional[List[Dict]] = None,
+    ) -> List[Tuple[Document, float]]:
+        params = {
+            "queryVector": embedding,
+            "path": self._embedding_key,
+            "numCandidates": 100,
+            "limit": k,
+            "index": self._index_name,
+        }
+        if pre_filter:
+            params["filter"] = pre_filter
+        query = {"$vectorSearch": params}
+
+        return self._similarity_search_query(query, post_filter_pipeline)
+
+    def _similarity_search_with_score(
+        self,
+        embedding: List[float],
+        k: int = 4,
+        pre_filter: Optional[dict] = None,
+        post_filter_pipeline: Optional[List[Dict]] = None,
+    ) -> List[Tuple[Document, float]]:
+        if self._use_vectorsearch:
+            try:
+                result = self._similarity_search_query_vectorsearch(
+                    embedding, k, pre_filter, post_filter_pipeline
+                )
+            except OperationFailure as e:
+                if e.code == 224:
```

Comment:
Could you add a comment here like "QueryFeatureNotAllowed error code"? Found here: https://github.com/mongodb/mongo/blob/0744ea1/src/mongo/base/error_codes.yml#L269C24-L269C46

Could you also confirm that this error code is stable for this use case?

Reply:
Will do, thanks!

Reply:
It appears the error code is specific to 6.0.10. Older server versions could give a variety of errors.